Move scripts from dev/ into ivadomed/scripts

charleygros opened the pull request:
### Aim
Move some scripts from `dev/` into `ivadomed/scripts` and document them on RTD.

Fixes #285.

jcohenadad commented on docs/source/scripts.rst:
```diff
@@ -0,0 +1,56 @@
+Utility scripts
+===============
+
+Visualize data augmentation transformations
```
not sure how they appear on the left TOC, but it is fairly small, so i would keep titles short, e.g. "Visualize transformations" or "Visualize augmentation transfos"
but this is a very minor detail...

jcohenadad replied:
update: after looking at the generated built doc, it looks fine the way it is 👍

jcohenadad commented on docs/source/scripts.rst:
```diff
@@ -0,0 +1,56 @@
+
+    python scripts/visualize_transforms.py -i t2s.nii.gz -n 1 -c config.json -r t2s_seg.nii.gz
+
+Provides a visualisation of a series of three transformation on a randomly selected slice:
```
```suggestion
Provides a visualization of a series of three transformation on a randomly selected slice:
```

jcohenadad commented on docs/source/scripts.rst:
```diff
@@ -0,0 +1,56 @@
+    :align: center
+    :height: 100px
+
+Convert a model from PyTorch to ONNX format
```
same here about length, how about "Convert to ONNX"
but the more i think about it, the more i like the "verbose" approach. Not sure how it will appear on the TOC. we could ask RTD to build the doc for this branch. i'll do it now...

jcohenadad commented on ivadomed/scripts/visualize_transforms.py:
```diff
@@ -12,7 +12,7 @@
 #   and save the visual result in a output folder o: transform0_slice19.png, transform0_transform1_slice19.png etc.
 #
-# Usage: python dev/visualize_transforms.py -i <input_filename> -c <fname_config> -n <int> -o <output_folder>
+# Usage: python scripts/visualize_transforms.py -i <input_filename> -c <fname_config> -n <int> -o <output_folder>
```
we should definitely add a script entry point in the setup.py, so we can write this instead:
```suggestion
# Usage: visualize_transforms -i <input_filename> -c <fname_config> -n <int> -o <output_folder>
```

jcohenadad commented:
i've added a build for this branch specifically: https://ivadomed.org/en/cg-scripts_folder/

jcohenadad commented on docs/source/scripts.rst:
```diff
@@ -0,0 +1,56 @@
+.. image:: ../../images/transforms_im.png
+    :width: 200px
+    :align: center
+    :height: 100px
```
why specifying both width and height? you want to keep the ratio
```suggestion
```

jcohenadad commented on docs/source/scripts.rst:
```diff
@@ -0,0 +1,56 @@
+Provides a visualisation of a series of three transformation on a randomly selected slice:
+
+.. image:: ../../images/transforms_im.png
+    :width: 200px
```
this needs to be much bigger. let's try this:
```suggestion
    :width: 600px
```

jcohenadad commented on docs/source/scripts.rst:
```diff
@@ -0,0 +1,56 @@
+.. image:: ../../images/transforms_gt.png
+    :width: 200px
+    :align: center
+    :height: 100px
```
```suggestion
```

jcohenadad commented on docs/source/scripts.rst:
```diff
@@ -0,0 +1,56 @@
+Gives:
+
+.. image:: ../../images/transforms_gt.png
+    :width: 200px
```
```suggestion
    :width: 600px
```

jcohenadad commented on docs/source/scripts.rst:
```diff
@@ -0,0 +1,56 @@
+
+    python scripts/convert_to_onnx.py -m path/to/model.pt -d 3
+
+Model's hyper-parameters optimisation
```
i more often see hyperparameter as one word-- and also to keep US convention (z)
```suggestion
Hyperparameter optimization
```

jcohenadad commented on docs/source/scripts.rst:
```diff
@@ -0,0 +1,56 @@
+
+Model's hyper-parameters optimisation
+***********************************************
+Model's hyper-parameters optimisation is tedious and time-consuming. ``scripts/automate_training.py`` allows to automatize this optimisation on multiple GPUs. The scripts runs trainings, on the same training and validation datasets, by combinating a given set of parameters and set of values for each of these parameters. Results are collected for each combination and reported into a dataframe to a ...
```
```suggestion
Hyperparameter optimization of models is tedious and time-consuming. ``scripts/automate_training.py`` automatizes this optimization across multiple GPUs. The scripts runs trainings, on the same training and validation datasets, by combining a given set of parameters and set of values for each of these parameters. Results are collected for each combination and reported into a dataframe to allow their comparison. The script efficiently allocates each training to one of the available GPUs. ::
```

jcohenadad commented on docs/source/scripts.rst:
```diff
@@ -0,0 +1,56 @@
+
+.. TODO: add example of DF
+
+Compare models' performance at inference time
```
```suggestion
Compare models performance at inference time
```

jcohenadad commented:
@charleygros i thought about something: instead of creating (and maintaining) a scripts.rst file, why not simply embedding all the material that you've already created inside the docstrings of each script, and use sphinx's automatic API documentation to generate the script utility?  Exactly like how Christian illustrated his scripts [here](https://medicaltorch.readthedocs.io/en/stable/modules.html#module-medicaltorch.models), using these [rich docstrings](https://medicaltorch.readthedocs.io/en/stable/_modules/medicaltorch/models.html#NoPoolASPP)

charleygros commented:
> @charleygros i thought about something: instead of creating (and maintaining) a scripts.rst file, why not simply embedding all the material that you've already created inside the docstrings of each script, and use sphinx's automatic API documentation to generate the script utility? Exactly like how Christian illustrated his scripts [here](https://medicaltorch.readthedocs.io/en/stable/modules.html#module-medicaltorch.models), using these [rich docstrings](https://medicaltorch.readthedocs.io/en/stable/_modules/medicaltorch/models.html#NoPoolASPP)

This is truly brilliant. I just did not know it was possible :-)

So for instance, we would get ride of "models.rst" as it would be automatically generated via sphinx's automatic API documentation.

For the `scripts`: We have only functions (not classes) --> would it work too? or is it using the docstring at the top of the script? I can also have a look

jcohenadad commented:
> So for instance, we would add network images to ivadomed/models.py docstrings and get ride of "models.rst" as it would be automatically generated via sphinx's automatic API documentation.

Yup!

> For the scripts: We have only functions (not classes) --> would it work too? or is it using the docstring at the top of the script? I can also have a look

Absolutely! In the rst file we just fetch the object with `automethod` instead of `autoclass`. You can see examples in the api_ref.rst

charleygros commented:
@jcohenadad : There is something I doing wrong.. not too sure what:
https://ivadomed.org/en/cg-scripts_folder/api_ref.html#ivadomed-scripts

Do you have an idea?

jcohenadad commented:
there is a problem in the way testing.utils is called, leading to the following error:
~~~
(venv-ivadomed) julien-macbook:~/code/ivadomed/docs $ sphinx-build source/ build
Running Sphinx v3.0.4

Configuration error:
There is a programmable error in your configuration file:

Traceback (most recent call last):
  File "/Users/julien/miniconda3/envs/venv-ivadomed/lib/python3.6/site-packages/sphinx/config.py", line 319, in eval_config_file
    execfile_(filename, namespace)
  File "/Users/julien/miniconda3/envs/venv-ivadomed/lib/python3.6/site-packages/sphinx/util/pycompat.py", line 81, in execfile_
    exec(code, _globals)
  File "/Users/julien/code/ivadomed/docs/source/conf.py", line 34, in <module>
    import ivadomed.scripts.visualize_transforms
  File "/Users/julien/code/ivadomed/ivadomed/scripts/visualize_transforms.py", line 13, in <module>
    from testing.utils import plot_transformed_sample
ModuleNotFoundError: No module named 'testing.utils'; 'testing' is not a package
~~~

the problem is that testing/ is not inside the ivadomed package (and does not have an __init__.py file in it). So either:
- we add a `sys.path.append()` in the scripts before importing from testing
- we move testing/ under ivadomed pkg for more convenient imports (which would make sense, given the dependency of testing.utils from some of the package methods).

jcohenadad commented:
same problem as above for non-importable APIs from dev/ folder:
~~~
(venv-ivadomed) julien-macbook:~/code/ivadomed/docs $ sphinx-build source/ build
Running Sphinx v3.0.4

Configuration error:
There is a programmable error in your configuration file:

Traceback (most recent call last):
  File "/Users/julien/miniconda3/envs/venv-ivadomed/lib/python3.6/site-packages/sphinx/config.py", line 319, in eval_config_file
    execfile_(filename, namespace)
  File "/Users/julien/miniconda3/envs/venv-ivadomed/lib/python3.6/site-packages/sphinx/util/pycompat.py", line 81, in execfile_
    exec(code, _globals)
  File "/Users/julien/code/ivadomed/docs/source/conf.py", line 36, in <module>
    import ivadomed.scripts.automate_training
  File "/Users/julien/code/ivadomed/ivadomed/scripts/automate_training.py", line 25, in <module>
    from dev.compare_models import compute_statistics
ModuleNotFoundError: No module named 'dev.compare_models'
~~~
for now, i will comment these imports

jcohenadad commented:
@charleygros the built doc is now exposing properly the methods in `scripts/`, see: https://ivadomed.org/en/cg-scripts_folder/scripts.html 

I've mentioned the "tricks" in the message of my latest commit 099d381.

charleygros commented:
Many thanks @jcohenadad for disentangling this!

Regarding the two imports that were causing problems:
`from testing.utils import plot_transformed_sample` --> this file (`testing.utils`) only contains this function, which could be actually of interest for other parts of the ivadomed code base --> I moved this function to `ivadomed.utils`, corrected the imports and removed `testing.utils`. It seems to me it makes more sense that way.
`from dev.compare_models import compute_statistics` was actually an error.. --> should be `from ivadomed.scripts.compare_models import compute_statistics` --> corrected

jcohenadad commented:
@charleygros awesome! i let you finish the inclusion of the rich docstrings in the other scripts? thanks!

jcohenadad commented:
> @charleygros awesome! i let you finish the inclusion of the rich docstrings in the other scripts? thanks!

nevermind i thought some scripts were still WIP
